src/core/auth/Auth.tsx

```diff
@@ -0,0 +1,113 @@
+import React, { createContext, ReactNode, useContext } from 'react';
+import { withRouter, RouteComponentProps } from 'react-router';
+
+import { getAuthorizedUser, LogInPayload, logIn, logOut, User } from 'api';
+
+import { Alert } from 'ui';
+
+namespace Auth {
+  export namespace Provider {
+    export interface State {
+      user: User;
+      pending: boolean;
+      authorized: boolean;
+      error: string;
+      logIn?(payload: LogInPayload): Promise<void>;
+      logOut?(): Promise<void>;
+    }
+
+    export interface Props extends RouteComponentProps {
+      children: ReactNode;
+    }
+  }
+}
+
+const STATE: Auth.Provider.State = {
+  user: null,
+  pending: false,
+  authorized: false,
+  error: ''
+};
+
+const Context = createContext(STATE);
+
+class Provider extends React.Component<Auth.Provider.Props, typeof STATE> {
+  componentDidMount() {
```

Comment:
typeof STATE?

Reply:
Y. It allows me to Save some boilerplate and Also i declared shape of state object only once and using this defnition in component. After possible refactor ot allows me easoer change models. Doing it only in one place.